File: python/test_solution.py
import pytest
from solution import has_cycle


def test_three_node_cycle():
    edges = [("a", "b"), ("b", "c"), ("c", "a")]
    assert has_cycle(edges) is True


def test_diamond_dag_no_cycle():
    edges = [("a", "b"), ("a", "c"), ("b", "d"), ("c", "d")]
    assert has_cycle(edges) is False


def test_self_loop():
    edges = [("a", "a")]
    assert has_cycle(edges) is True


def test_empty_edges():
    edges = []
    assert has_cycle(edges) is False


def test_disjoint_components_one_has_cycle():
    edges = [("x", "y"), ("y", "z"), ("p", "q"), ("q", "p")]
    assert has_cycle(edges) is True

File: cpp/test_solution.cpp
#include "solution.h"
#include <cassert>
#include <iostream>
#include <string>
#include <vector>
#include <utility>

int main() {
    // Three-node cycle
    {
        std::vector<std::pair<std::string,std::string>> edges = {
            {"a","b"}, {"b","c"}, {"c","a"}
        };
        assert(hasCycle(edges) == true);
    }

    // Diamond DAG — no cycle
    {
        std::vector<std::pair<std::string,std::string>> edges = {
            {"a","b"}, {"a","c"}, {"b","d"}, {"c","d"}
        };
        assert(hasCycle(edges) == false);
    }

    // Self-loop
    {
        std::vector<std::pair<std::string,std::string>> edges = {
            {"a","a"}
        };
        assert(hasCycle(edges) == true);
    }

    // Empty edge list
    {
        std::vector<std::pair<std::string,std::string>> edges = {};
        assert(hasCycle(edges) == false);
    }

    // Disjoint components — only one has a cycle
    {
        std::vector<std::pair<std::string,std::string>> edges = {
            {"x","y"}, {"y","z"}, {"p","q"}, {"q","p"}
        };
        assert(hasCycle(edges) == true);
    }

    std::cout << "All tests passed" << std::endl;
    return 0;
}
